tracer/src/Datadog.Trace/Span.cs

```diff
@@ -336,6 +336,7 @@ internal void Finish(TimeSpan duration)
             lock (_lock)
             {
                 ResourceName ??= OperationName;
+                Context.TraceContext?.IastRequestContext?.AddIastTagsIfNeeded(this);
```

Comment:
Usually when we want to add tags to a span after it is closed, we do it inside `TraceContext.CloseSpan()`. I don't have a strong argument aside from keeping all the tagging code together where it's easy to find.

Reply:
The reason I was not doing it in tracecontext.close() is that the span class does not seem to allow setting tags if IsFinished is true, which is set before calling tracecontext.close. Still, I agree, if there is a way to add this code to tracecontext.close(), it would be better.

Reply:
> the span class does not seem to allow setting tags if `IsFinished` is true

Ah, yes. Instead of the public `span.SetTag()`, you can use the internal `span.Tags.SetTag()` for this. I wasn't aware of this inconsistency until Robert brought it up recently.

We could also add the tags during MessagePack serialization, which would have the benefits of moving the JSON serialization into the serialization thread, but we can do that later.

Reply:
We need to be quite careful here. Using the internal `span.Tags.SetTag()` to set tags after close can create a race. The tags might be set after the span is already serialized. I guess that might not happen in this specific case, if the root span is still open?

Reply:
Moving that call to the locked part of the method Finish in span.cs would prevent the race condition which is not so clean, but would prevent that race condition for sure. In requests, we only add the tags to the root span and now it is done right after setting it's ItsFinished property to true, in the tracecontext CloseSpan method. There has been some discussion regarding the specific place where this call should be done, but avoiding the race condition would be a strong argument to put the call back to span.cs. I can do that easily if we agree that this place could be a better option.

Reply:
What do you think, @lucaspimentel?

Reply:
Anywhere inside `TraceContext.CloseSpan(Span)` should be safe as long as it's before the `Tracer.Write(spansToWrite)` call at the end.

Reply:
Also, there's already an
```
if (span.IsRootSpan && span.Type == SpanTypes.Web)
```
which IIRC is what you want, so you could do it in there.